libraries/botbuilder-dialogs-adaptive-runtime/src/coreBotAdapter.ts

```diff
@@ -1,50 +1,76 @@
 // Copyright (c) Microsoft Corporation.
 // Licensed under the MIT License.
 
-import { ClaimsIdentity, SkillValidation } from 'botframework-connector';
+import * as t from 'runtypes';
+import { BotFrameworkAuthentication, ClaimsIdentity, SkillValidation } from 'botframework-connector';
 
 import {
     ActivityEx,
-    BotFrameworkAdapter,
-    BotFrameworkAdapterSettings,
+    CloudAdapter,
     ConversationState,
     InputHints,
     MessageFactory,
+    Response,
     TurnContext,
     UserState,
     useBotState,
 } from 'botbuilder';
 
-export class CoreBotAdapter extends BotFrameworkAdapter {
+const ErrorT = t.Record({
+    message: t.String,
+    statusCode: t.Optional(t.Number),
+});
+
+export class CoreBotAdapter extends CloudAdapter {
     constructor(
-        settings: Partial<BotFrameworkAdapterSettings>,
+        botFrameworkAuthentication: BotFrameworkAuthentication,
         private readonly conversationState: ConversationState,
         userState: UserState
     ) {
-        super(settings);
+        super(botFrameworkAuthentication);
 
         useBotState(this, userState, conversationState);
 
-        this.onTurnError = async (turnContext, err) => {
+        this.onTurnError = async (context, err) => {
             console.error('[onTurnError] unhandled error', err);
 
-            await this.sendErrorMessage(turnContext, err);
-            await this.sendEoCToParentIfSkill(turnContext, err);
-            await this.clearConversationState(turnContext);
+            const res = context.turnState.get<Response | undefined>(this.ResponseKey);
+
+            if (res) {
+                // Try to extract errors wrapped in, for example, DialogContextError
+                let anyErr = err as any; // eslint-disable-line @typescript-eslint/no-explicit-any
+                if (anyErr.error instanceof Error) {
+                    anyErr = anyErr.error;
+                }
+
+                if (ErrorT.guard(anyErr)) {
+                    res.status(anyErr.statusCode ?? 500);
+                    res.send(anyErr.message);
+                } else {
+                    res.status(500);
+                    res.send(anyErr);
+                }
+
+                res.end();
+            }
+
+            await this.sendErrorMessage(context, err);
+            await this.sendEoCToParentIfSkill(context, err);
+            await this.clearConversationState(context);
         };
     }
 
-    private async sendErrorMessage(turnContext: TurnContext, error: Error): Promise<void> {
+    private async sendErrorMessage(context: TurnContext, error: Error): Promise<void> {
```

Comment:
Should we make the default error message configurable (`errorMessageText`)? I could see developers wanting to customize that before publishing, which I think would be difficult to do inside of Composer, as-is.

Reply:
At the moment that isn't supported by the .NET runtime. I do intend to reconcile differences with the current [CoreBotAdapter](https://github.com/microsoft/botbuilder-dotnet/blob/main/libraries/Microsoft.Bot.Builder.Dialogs.Adaptive.Runtime/CoreBotAdapter.cs), though.